instrumentation/kafka-clients/src/main/java/brave/kafka/clients/TracingProducer.java

```diff
@@ -59,10 +63,24 @@
     return this.send(record, null);
   }
 
-  /** We wrap the send method to add tracing. */
-  @Override
-  public Future<RecordMetadata> send(ProducerRecord<K, V> record, @Nullable Callback callback) {
-    Span span = tracing.tracer().nextSpan();
+  /**
+   * We wrap the send method to add tracing.
+   * This method is expecting an existing current trace context to create a child span.
+   * If there is no current context, an additional operation will be performed to extract
+   * context from record headers. This could impact performance if no context is injected
+   * on record headers, but is required to propagate context on Kafka Streams instrumentation.
+   */
+  @Override public Future<RecordMetadata> send(ProducerRecord<K, V> record, @Nullable Callback callback) {
+    final Span span;
+    final TraceContext maybeParent = tracing.currentTraceContext().get();
```

Comment:
the key here is that the order is flipped. for example, usually we try to extract and then if that fails use the current span. Here we are looking for the current span first (because we are on the sender side!)